docs: Add updated component integration guide (#1488)
opendatahub-io/opendatahub-operator: docs/COMPONENT_INTEGRATION.md

```go
func (r *DataScienceClusterReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager) error {
	componentsPredicate := dependent.New(dependent.WithWatchStatus(true))

	return ctrl.NewControllerManagedBy(mgr).
		For(&dscv1.DataScienceCluster{}, builder.WithPredicates(predicates.DefaultPredicate)).
		// components
		Owns(&componentApi.Dashboard{}, builder.WithPredicates(componentsPredicate)).
		// ... other existing integrated components ...
		Owns(&componentApi.ExampleComponent{}, builder.WithPredicates(componentsPredicate)).
		Watches(
			&dsciv1.DSCInitialization{},
			handlers.Fn(r.watchDataScienceClusters)).
		Watches(
			&corev1.ConfigMap{},
			handlers.Fn(r.watchDataScienceClusters),
			builder.WithPredicates(r.filterDeleteConfigMap())).
		Complete(r)
}
```

#### Update upgrade.go

Update the `CreateDefaultDSC()` function to include the newly added component.

#### Update main.go

Add an import for the the newly added component:

```go
package main

import (
	// ... existing imports ...

	// ... component imports for the integrated components ...
	_ "github.com/opendatahub-io/opendatahub-operator/v2/controllers/components/<example_component>"
)
```

> 

Author: Added a section for adding kubebuilder RBAC rules, thank you

### 3. Add unit and e2e tests

Please add `unit` tests for any component-specific functions added to the codebase.

In addition, please add [e2e tests](https://github.com/opendatahub-io/opendatahub-operator/tree/main/tests/e2e) to
the e2e test suite to capture deployments introduced by the new component.

> 

Author: Good catch, thank you, updated

## Integrating a new component

To ensure a new component is integrated seamlessly in the operator, please follow the steps listed below.

### 1. Update API specs

The first step is to define the internal API spec for the new component and introduce it to the existing DataScienceCluster (DSC) API. Please proceed as the following:

#### Define internal API spec for the new component

1. Create a dedicated `<example_component_name>_types.go` file within `apis/components/v1alpha1` directory.

2. Define the internal API spec for the new component according to the expected definitions.
You can use the following pseudo-implementation for reference:

```go
const (
  	// example new component name
	ExampleComponentName         = "example-component"

	// ExampleComponentInstanceName is the name of the new component instance singleton
	// value should match what is set in the kubebuilder markers for XValidation defined below
	ExampleComponentInstanceName = "default-example-component"

  	// kubernetes kind of the new component
	ExampleComponentKind         = "ExampleComponent"
)

type ExampleComponentCommonSpec struct {
	// new component spec exposed to DSC api

> 	components.DevFlagsSpec `json:",inline"`

Reviewer: this will be I think `common.DevFlagsSpec `json:",inline"`` similarly other places

Reviewer: "github.com/opendatahub-io/opendatahub-operator/v2/apis/common"

For practical examples of all the abovementioned functionality, please refer to the implementations within `controllers/components` directory.

#### Register new component to the manager (DSC controller)

Register the new component API to the manager as the following pseudo-implementation:
```go
func (r *DataScienceClusterReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager) error {
	componentsPredicate := dependent.New(dependent.WithWatchStatus(true))

	return ctrl.NewControllerManagedBy(mgr).
		For(&dscv1.DataScienceCluster{}, builder.WithPredicates(predicates.DefaultPredicate)).
		// components
		Owns(&componentApi.Dashboard{}, builder.WithPredicates(componentsPredicate)).
		// ... other existing integrated components ...
		Owns(&componentApi.ExampleComponent{}, builder.WithPredicates(componentsPredicate)).
		Watches(
			&dsciv1.DSCInitialization{},
			handlers.Fn(r.watchDataScienceClusters)).
		Watches(
			&corev1.ConfigMap{},
			handlers.Fn(r.watchDataScienceClusters),
			builder.WithPredicates(r.filterDeleteConfigMap())).
		Complete(r)
}
```

#### Update upgrade.go

> Update the `CreateDefaultDSC()` function to include the newly added component.

Reviewer: maybe add location of upgrade.go

- `initialize()` - to register paths to the component manifests
- `devFlags()` - to override the component manifest paths according to the Dev Flags configuration 

In addition, proper generic actions, intended to be used across the components, are provided as part of the operator implementation (located in `pkg/controller/actions`).
These support:
- (if necessary) creating pod security role binding
- manifest rendering
    - can additionally utilize caching
- manifest deployment
    - can additionally utilize caching
- status updating
- garbage collection
	- **additional requirement - garbage collection action must always be called as the last action before the final `.Build()` call**

In case the new component requires additional custom logic, custom actions can also be added to the builder via the respective `.WithAction()` calls.

For practical examples of all the abovementioned functionality, please refer to the implementations within `controllers/components` directory.

#### Register new component to the manager (DSC controller)

Register the new component API to the manager as the following pseudo-implementation:

> ```go

Reviewer 2: ```suggestion
```diff
```

Using a `diff` code block rather than a `go` code block like this will allow highlighting the line to be added, at the expense of language-specific syntax highlighting. I think it's a worthwhile tradeoff in this case (there might be other cases like that here where it might be applicable too).

> func (r *DataScienceClusterReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager) error {

Reviewer 3: this is not required anymore, as long as the component is [registered to the component registry ](https://github.com/opendatahub-io/opendatahub-operator/blob/4697aacd6e64005145d0ce158c7f8013180f775e/controllers/components/codeflare/codeflare.go#L26-L28), then is is automatically added to the manager

	componentsPredicate := dependent.New(dependent.WithWatchStatus(true))

	return ctrl.NewControllerManagedBy(mgr).
		For(&dscv1.DataScienceCluster{}, builder.WithPredicates(predicates.DefaultPredicate)).
		// components
		Owns(&componentApi.Dashboard{}, builder.WithPredicates(componentsPredicate)).
		// ... other existing integrated components ...

> 		Owns(&componentApi.ExampleComponent{}, builder.WithPredicates(componentsPredicate)).

Reviewer 2: ```suggestion
+		Owns(&componentApi.ExampleComponent{}, builder.WithPredicates(componentsPredicate)).
```

If you decide to use the `diff` code block suggestion from above, then using a `+` here for the newly added line will highlight it in green in the rendered block, making it clearer what is expected to be changed in this section.

Author: Thank you Gerard, this section was removed based on Luca's feedback, but I introduced it in code examples where structs/imports get extended.

#### Update main.go

Add an import for the the newly added component:

```go
package main

import (
	// ... existing imports ...

	// ... component imports for the integrated components ...
	_ "github.com/opendatahub-io/opendatahub-operator/v2/controllers/components/<example_component>"
)
```

### 3. Add unit and e2e tests

Please add `unit` tests for any component-specific functions added to the codebase.

Please also add [e2e tests](https://github.com/opendatahub-io/opendatahub-operator/tree/main/tests/e2e) to
the e2e test suite to capture deployments introduced by the new component.
Existing e2e test suites for the integrated components can be also found there.

Lastly, please update the following files to fully integrate new component tests into the overall test suite:
- update `setupDSCInstance()` function in `tests/e2e/helper_test.go` to include the newly added component
- update `newDSC()` function in `controllers/webhook/webhook_suite_test.go` to include the newly added component
- update `componentsTestSuites` map in `tests/e2e/controller_test.go` to include the reference for the new component e2e test suite

## Integrated components

> Currently integrated components are:

Reviewer 2: Could optionally add ModelController to this list also now.

- [Dashboard](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/dashboard)
- [Codeflare](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/codeflare)
- [Ray](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/ray)
- [Data Science Pipelines](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/datasciencepipelines)
- [KServe](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/kserve)
- [ModelMesh Serving](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/modelmeshserving)
- [Workbenches](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/workbenches)
- [TrustyAI](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/trustyai)
- [ModelRegistry](https://github.com/opendatahub-io/opendatahub-operator/tree/main/components/modelregistry)
- [Kueue](https://github.com/opendatahub-io/kueue)

> The particular controller implementations for the listed components are located in `controllers/components` directory and the corresponding internal component APIs are located in `apis/components/v1alpha1`.

Reviewer 2: ```suggestion
The particular controller implementations for the listed components are located in the `controllers/components` directory and the corresponding internal component APIs are located in `apis/components/v1alpha1`.
```

# Component Integration

Since the ODH operator is the integration point to deploy ODH component manifests, it is essential to have common processes to integrate new components.

Currently, each component is expected to have its own dedicated internal API/CRD and dedicated reconciler.
To understand the current operator architecture and its inner workings, please refer to [the design document](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/DESIGN.md).

The list of the currently integrated ODH components is provided [at the end of this document](#integrated-components).

## Integrating a new component

To ensure a new component is integrated seamlessly in the operator, please follow the steps listed below.

### 1. Update API specs

> The first step is to define the internal API spec for the new component and introduce it to the existing DataScienceCluster (DSC) API. Please proceed as the following:

Reviewer 2: ```suggestion
The first step is to define the internal API spec for the new component and introduce it to the existing DataScienceCluster (DSC) API. Please proceed as follows:
```

As seen in the existing component reconciler implementations, it would be recommended to include the action implementations in a separate file within the module, such as `<example_component_name>_controller_actions.go`.

"Generic"/commonly-implemented actions for each of the currently integrated components include:
- `initialize()` - to register paths to the component manifests
- `devFlags()` - to override the component manifest paths according to the Dev Flags configuration 

In addition, proper generic actions, intended to be used across the components, are provided as part of the operator implementation (located in `pkg/controller/actions`).
These support:
- (if necessary) creating pod security role binding
- manifest rendering
    - can additionally utilize caching
- manifest deployment
    - can additionally utilize caching
- status updating
- garbage collection
	- **additional requirement - garbage collection action must always be called as the last action before the final `.Build()` call**

If the new component requires additional custom logic, custom actions can also be added to the builder via the respective `.WithAction()` calls.

> For practical examples of all the abovementioned functionality, please refer to the implementations within `controllers/components` directory.

Reviewer 2: ```suggestion
For practical examples of all the above-mentioned functionality, please refer to the implementations within `controllers/components` directory.
```
